source/agora/common/BitField.d

```diff
@@ -23,117 +34,173 @@ import agora.common.Hash;
-        this.num_bits = num_bits;
-        this.bit_array.initialize(this.storage, num_bits);
+        this._storage = new inout(T)[(min_num_bits + BitsPerT - 1) / BitsPerT];
     }
 
+    @nogc:
+
     /***************************************************************************
 
-        Serialization. The data is packed tightly by
-        removing unused bytes from the uint[] storage,
-        and the length of bits is encoded instead of
-        the length of bytes.
+        Build a `BitField` from an existing storage
 
         Params:
-            dg = serialize function accumulator
+          storage = Backing store to use for BitField.
 
     ***************************************************************************/
 
-    public void serialize (scope SerializeDg dg) const @trusted
+    public this (inout(T)[] storage) inout
     {
-        serializePart(this.num_bits, dg);
-        serializePart(this.storage, dg);
+        this._storage = storage;
     }
 
-    /***************************************************************************
-
-        Deserialization.
-
-        Params:
-            dg = deserialize function accumulator
-
-    ***************************************************************************/
-
-    public void deserialize (scope DeserializeDg dg) @safe
+    /// Returns the number of bits in the bitfield
+    public size_t length () const
     {
-        uint num_bits;
-        deserializePart(num_bits, dg);
-
-        deserializePart(this.storage, dg);
-        () @trusted { this.bit_array.initialize(this.storage, num_bits); }();
-        this.num_bits = num_bits;
+        return this._storage.length * BitsPerT;
     }
 
-    /// Custom equality support
-    public int opEquals ( in BitField rhs_ ) const nothrow @trusted
+    /// Supports setting a single bit to a value
+    public bool opIndexAssign (bool value, size_t index)
     {
-        scope (failure) assert(0);  // BitArray lacks nothrow support
+        if (index >= this.length())
+            assert(0);
+
+        if (value)
+            this._storage[index / BitsPerT] |= mask(index);
+        else
+            // This line triggers an annoying deprecation which we can't
+            // seem to get rid of. TODO: Report to DMD
+            this._storage[index / BitsPerT] &= ~mask(index);
+        return value;
+    }
 
-        // cast: BitArray does not support const yet
-        auto lhs = cast()this.bit_array;
-        auto rhs = cast()rhs_.bit_array;
-        return lhs.opEquals(rhs);
+    /// Gets a single bit's value
+    public bool opIndex (size_t index)
+    {
+        if (index >= this.length())
+            assert(0);
+        return !!(this._storage[index / BitsPerT] & mask(index));
     }
 
-    /// Hashing support
-    public void computeHash (scope HashDg dg) const nothrow @nogc
+    /// Compare BitField of same length
+    public bool opEquals (OtherT) (const auto ref BitField!OtherT other) const
```

Comment:
I think there is a unittest missing for comparing two different backing-store Bitfield types?

Reply:
Line 57 (and after) has it. `bf3` uses `ubyte` as backing store, `bf2` uses `uint`.